codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/endpoint/generators/EndpointParamsInterceptorGenerator.kt

```diff
@@ -0,0 +1,147 @@
+    private val codegenScope = codegenContext.runtimeConfig.let { rc ->
+        val endpointTypesGenerator = EndpointTypesGenerator.fromContext(codegenContext)
+        val runtimeApi = CargoDependency.smithyRuntimeApi(rc).toType()
+        val interceptors = runtimeApi.resolve("client::interceptors")
+        val orchestrator = runtimeApi.resolve("client::orchestrator")
+        arrayOf(
+            "BoxError" to runtimeApi.resolve("client::runtime_plugin::BoxError"),
+            "ConfigBag" to runtimeApi.resolve("config_bag::ConfigBag"),
+            "EndpointResolverParams" to orchestrator.resolve("EndpointResolverParams"),
+            "HttpResponse" to orchestrator.resolve("HttpResponse"),
+            "HttpRequest" to orchestrator.resolve("HttpRequest"),
+            "Interceptor" to interceptors.resolve("Interceptor"),
+            "InterceptorContext" to interceptors.resolve("InterceptorContext"),
+            "InterceptorError" to interceptors.resolve("error::InterceptorError"),
+            "ParamsBuilder" to endpointTypesGenerator.paramsBuilder(),
+        )
+    }
+
+    fun render(writer: RustWriter, operationShape: OperationShape) {
+        val operationName = symbolProvider.toSymbol(operationShape).name
+        renderInterceptor(
+            writer,
+            "${operationName}EndpointParamsInterceptor",
+            implInterceptorBodyForEndpointParams(operationShape),
+        )
+        renderInterceptor(
+            writer, "${operationName}EndpointParamsFinalizerInterceptor",
+            implInterceptorBodyForEndpointParamsFinalizer,
+        )
+    }
+
+    private fun renderInterceptor(writer: RustWriter, interceptorName: String, implBody: Writable) {
+        writer.rustTemplate(
+            """
+            ##[derive(Debug)]
+            struct $interceptorName;
+
+            impl #{Interceptor}<#{HttpRequest}, #{HttpResponse}> for $interceptorName {
+                fn read_before_execution(
+                    &self,
+                    context: &#{InterceptorContext}<#{HttpRequest}, #{HttpResponse}>,
+                    cfg: &mut #{ConfigBag},
+                ) -> Result<(), #{BoxError}> {
+                    #{body:W}
+                }
+            }
+            """,
+            *codegenScope,
+            "body" to implBody,
+        )
+    }
+
+    private fun implInterceptorBodyForEndpointParams(operationShape: OperationShape): Writable = writable {
+        val operationInput = symbolProvider.toSymbol(operationShape.inputShape(model))
+        rustTemplate(
+            """
+            let input = context.input()?;
+            let _input = input
+                .downcast_ref::<${operationInput.name}>()
+                .ok_or_else(|| #{InterceptorError}::invalid_input_access())?;
+            let params_builder = cfg
+                .get::<#{ParamsBuilder}>()
+                .ok_or(#{InterceptorError}::read_before_execution("missing endpoint params builder"))?
```

Comment:
Should be `.ok_or_else(|| "missing endpoint params builder".to_string())?`

Reply:
Addressed in 10a1967.